File: installers/image_viewer.py
# SPDX-FileCopyrightText: Copyright (C) 2025 Akshat Kotpalliwar (alias IntegerAlex) <inquiry.akshatkotpalliwar@gmail.com>
# SPDX-License-Identifier: GPL-3.0-only
"""Advance Image Viewer installer."""

import shutil

from utils.base import BaseInstaller
from utils.types import InstallerResult


class ImageViewerInstaller(BaseInstaller):
    """Install Advance Image Viewer."""
    
    def is_installed(self) -> bool:
        # Check if command exists (common names for advance image viewer)
        if shutil.which("aiv") or shutil.which("advance-image-viewer"):
            return True
        # Check if installed via script location
        try:
            result = self.runner.run(["which", "aiv"], check=False, capture_output=True)
            return result.returncode == 0
        except Exception:
            return False
    
    def install(self) -> InstallerResult:
        try:
            self.console.print("[blue]ℹ[/blue] Installing Advance Image Viewer")
            
            install_script = "curl -fsSL https://advance-image-viewer.gossorg.in | bash"
            self.runner.run(
                ["bash", "-c", install_script],
                description="Installing Advance Image Viewer"
            )
            
            return InstallerResult(True, "Advance Image Viewer installed successfully")
        except Exception as e:
            return InstallerResult(False, "Failed to install Advance Image Viewer", str(e))

File: src/installers/image-viewer.ts
import type { CommandRunner } from "../utils/command-runner.js";
import type { SetupLogger } from "../utils/logger.js";
import type { InstallerResult } from "../types.js";
import { blue } from "../utils/ansi.js";

export function createImageViewerInstaller(runner: CommandRunner, logger: SetupLogger) {
  function checkInstalled(): boolean {
    try {
      const r = runner.run(["which", "aiv"], { check: false, capture: true });
      if ((r.status ?? 1) === 0) return true;
      const rr = runner.run(["which", "advance-image-viewer"], { check: false, capture: true });
      return (rr.status ?? 1) === 0;
    } catch {
      return false;
    }
  }

  async function doInstall(): Promise<InstallerResult> {
    try {
      process.stdout.write(`${blue("\u2139")} Installing Advance Image Viewer\n`);
      runner.run(
        ["bash", "-c", "curl -fsSL https://advance-image-viewer.gossorg.in | bash"],
        { description: "Installing Advance Image Viewer" },
      );
      return { success: true, message: "Advance Image Viewer installed successfully" };
    } catch (e) {
      return { success: false, message: "Failed to install Advance Image Viewer", error: String(e) };
    }
  }

  return { is_installed: checkInstalled, install: doInstall };
}
